src/datasets/table.py

```diff
@@ -1933,13 +1940,14 @@ def array_cast(array: pa.Array, pa_type: pa.DataType, allow_number_to_str=True):
             array_offsets = (np.arange(len(array) + 1) + array.offset) * array.type.list_size
             return pa.ListArray.from_arrays(array_offsets, _c(array.values, pa_type.value_type), mask=array.is_null())
     else:
-        if (
-            not allow_number_to_str
-            and pa.types.is_string(pa_type)
-            and (pa.types.is_floating(array.type) or pa.types.is_integer(array.type))
+        if pa.types.is_string(pa_type) and (
+            (not allow_primitive_to_str and pa.types.is_primitive(array.type))
+            or (not allow_decimal_to_str and pa.types.is_decimal(array.type))
         ):
             raise TypeError(
-                f"Couldn't cast array of type {array.type} to {pa_type} since allow_number_to_str is set to {allow_number_to_str}"
+                f"Couldn't cast array of type {array.type} to {pa_type} "
+                f"since allow_primitive_to_str is set to {allow_primitive_to_str} "
+                f"and allow_decimal_to_str is set to {allow_decimal_to_str}."
```

Comment:
Maybe let's be more precise here depending on the type - if it's primitive, then we should mention `allow_primitive_to_str `, and if it's decimal  `allow_decimal_to_str `

Reply:
Split this into two separate errors.